### ISSUE TITLE
ludwig_model.train for data_csv and data_df get different training results

### ISSUE THREAD

**Issue by danielduckworth** [uber/ludwig#483]
**Describe the bug**
When fitting a model with the API I get different results when using a dataframe compared to a csv.

When using the csv, everything is fine.

When using the dataframe, early stopping activates after 5 epochs even though there has objectively been an improvement to the loss.

**To Reproduce**
```
model.train(data_df = df_processed) #Not working as expected
model.train(data_csv = "data.csv") #Working as expected
```

**Expected behavior**
Fitting the model to the data in the dataframe should be similar to fitting data from the csv.

**Screenshots**
```
Epoch    1
INFO:ludwig.models.model:
Epoch    1
Training: 100%|██████████| 5/5 [00:01<00:00,  2.76it/s]
Evaluation train: 100%|██████████| 5/5 [00:00<00:00, 13.31it/s]
Evaluation vali : 100%|██████████| 1/1 [00:00<00:00, 43.32it/s]
Evaluation test : 100%|██████████| 2/2 [00:00<00:00, 36.57it/s]
Took 2.3470s
INFO:ludwig.models.model:Took 2.3470s
╒═════════╤════════════╤══════════════════════╤═══════════════════════╤═════════╤═════════╕
│ Score   │       loss │   mean_squared_error │   mean_absolute_error │      r2 │   error │
╞═════════╪════════════╪══════════════════════╪═══════════════════════╪═════════╪═════════╡
│ train   │ 10114.1474 │           10114.1474 │               99.2149 │ -0.3906 │ 99.2149 │
├─────────┼────────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ vali    │ 10252.2337 │           10252.2337 │               99.5787 │ -0.4162 │ 99.5787 │
├─────────┼────────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ test    │ 10189.9200 │           10189.9200 │               99.4607 │ -0.3820 │ 99.4607 │
╘═════════╧════════════╧══════════════════════╧═══════════════════════╧═════════╧═════════╛
INFO:ludwig.models.model:╒═════════╤════════════╤══════════════════════╤═══════════════════════╤═════════╤═════════╕
│ Score   │       loss │   mean_squared_error │   mean_absolute_error │      r2 │   error │
╞═════════╪════════════╪══════════════════════╪═══════════════════════╪═════════╪═════════╡
│ train   │ 10114.1474 │           10114.1474 │               99.2149 │ -0.3906 │ 99.2149 │
├─────────┼────────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ vali    │ 10252.2337 │           10252.2337 │               99.5787 │ -0.4162 │ 99.5787 │
├─────────┼────────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ test    │ 10189.9200 │           10189.9200 │               99.4607 │ -0.3820 │ 99.4607 │
╘═════════╧════════════╧══════════════════════╧═══════════════════════╧═════════╧═════════╛
Last improvement of loss on combined happened 1 epoch ago
INFO:ludwig.models.model:Last improvement of loss on combined happened 1 epoch ago

INFO:ludwig.models.model:

Epoch    2
INFO:ludwig.models.model:
Epoch    2
Training: 100%|██████████| 5/5 [00:00<00:00,  7.71it/s]
Evaluation train: 100%|██████████| 5/5 [00:00<00:00, 26.75it/s]
Evaluation vali : 100%|██████████| 1/1 [00:00<00:00, 39.34it/s]
Evaluation test : 100%|██████████| 2/2 [00:00<00:00, 34.57it/s]
Took 0.9422s
INFO:ludwig.models.model:Took 0.9422s
╒═════════╤══════════╤══════════════════════╤═══════════════════════╤════════╤═════════╕
│ Score   │     loss │   mean_squared_error │   mean_absolute_error │     r2 │   error │
╞═════════╪══════════╪══════════════════════╪═══════════════════════╪════════╪═════════╡
│ train   │ 279.4481 │             279.4481 │               14.1176 │ 0.0000 │  6.5321 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼────────┼─────────┤
│ vali    │ 312.4302 │             312.4302 │               14.7603 │ 0.0022 │  8.0211 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼────────┼─────────┤
│ test    │ 315.5273 │             315.5273 │               14.9357 │ 0.0010 │  6.9688 │
╘═════════╧══════════╧══════════════════════╧═══════════════════════╧════════╧═════════╛
INFO:ludwig.models.model:╒═════════╤══════════╤══════════════════════╤═══════════════════════╤════════╤═════════╕
│ Score   │     loss │   mean_squared_error │   mean_absolute_error │     r2 │   error │
╞═════════╪══════════╪══════════════════════╪═══════════════════════╪════════╪═════════╡
│ train   │ 279.4481 │             279.4481 │               14.1176 │ 0.0000 │  6.5321 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼────────┼─────────┤
│ vali    │ 312.4302 │             312.4302 │               14.7603 │ 0.0022 │  8.0211 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼────────┼─────────┤
│ test    │ 315.5273 │             315.5273 │               14.9357 │ 0.0010 │  6.9688 │
╘═════════╧══════════╧══════════════════════╧═══════════════════════╧════════╧═════════╛
Last improvement of loss on combined happened 2 epochs ago
INFO:ludwig.models.model:Last improvement of loss on combined happened 2 epochs ago

INFO:ludwig.models.model:

Epoch    3
INFO:ludwig.models.model:
Epoch    3
Training: 100%|██████████| 5/5 [00:00<00:00,  7.72it/s]
Evaluation train: 100%|██████████| 5/5 [00:00<00:00, 25.79it/s]
Evaluation vali : 100%|██████████| 1/1 [00:00<00:00, 41.65it/s]
Evaluation test : 100%|██████████| 2/2 [00:00<00:00, 33.68it/s]
Took 0.9449s
INFO:ludwig.models.model:Took 0.9449s
╒═════════╤═══════════╤══════════════════════╤═══════════════════════╤═════════╤══════════╕
│ Score   │      loss │   mean_squared_error │   mean_absolute_error │      r2 │    error │
╞═════════╪═══════════╪══════════════════════╪═══════════════════════╪═════════╪══════════╡
│ train   │ 1555.2219 │            1555.2219 │               36.7704 │ -0.0603 │ -36.7704 │
├─────────┼───────────┼──────────────────────┼───────────────────────┼─────────┼──────────┤
│ vali    │ 1416.2897 │            1416.2897 │               35.0317 │ -0.0442 │ -35.0260 │
├─────────┼───────────┼──────────────────────┼───────────────────────┼─────────┼──────────┤
│ test    │ 1557.9556 │            1557.9556 │               36.3586 │ -0.0530 │ -36.3261 │
╘═════════╧═══════════╧══════════════════════╧═══════════════════════╧═════════╧══════════╛
INFO:ludwig.models.model:╒═════════╤═══════════╤══════════════════════╤═══════════════════════╤═════════╤══════════╕
│ Score   │      loss │   mean_squared_error │   mean_absolute_error │      r2 │    error │
╞═════════╪═══════════╪══════════════════════╪═══════════════════════╪═════════╪══════════╡
│ train   │ 1555.2219 │            1555.2219 │               36.7704 │ -0.0603 │ -36.7704 │
├─────────┼───────────┼──────────────────────┼───────────────────────┼─────────┼──────────┤
│ vali    │ 1416.2897 │            1416.2897 │               35.0317 │ -0.0442 │ -35.0260 │
├─────────┼───────────┼──────────────────────┼───────────────────────┼─────────┼──────────┤
│ test    │ 1557.9556 │            1557.9556 │               36.3586 │ -0.0530 │ -36.3261 │
╘═════════╧═══════════╧══════════════════════╧═══════════════════════╧═════════╧══════════╛
Last improvement of loss on combined happened 3 epochs ago
INFO:ludwig.models.model:Last improvement of loss on combined happened 3 epochs ago

INFO:ludwig.models.model:

Epoch    4
INFO:ludwig.models.model:
Epoch    4
Training: 100%|██████████| 5/5 [00:00<00:00,  7.62it/s]
Evaluation train: 100%|██████████| 5/5 [00:00<00:00, 26.76it/s]
Evaluation vali : 100%|██████████| 1/1 [00:00<00:00, 40.81it/s]
Evaluation test : 100%|██████████| 2/2 [00:00<00:00, 35.42it/s]
Took 0.9476s
INFO:ludwig.models.model:Took 0.9476s
╒═════════╤══════════╤══════════════════════╤═══════════════════════╤═════════╤═════════╕
│ Score   │     loss │   mean_squared_error │   mean_absolute_error │      r2 │   error │
╞═════════╪══════════╪══════════════════════╪═══════════════════════╪═════════╪═════════╡
│ train   │ 793.5813 │             793.5813 │               25.1056 │ -0.0183 │ 24.5648 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ vali    │ 856.6062 │             856.6062 │               26.0861 │ -0.0207 │ 25.4445 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ test    │ 848.6282 │             848.6282 │               25.8354 │ -0.0184 │ 24.9674 │
╘═════════╧══════════╧══════════════════════╧═══════════════════════╧═════════╧═════════╛
INFO:ludwig.models.model:╒═════════╤══════════╤══════════════════════╤═══════════════════════╤═════════╤═════════╕
│ Score   │     loss │   mean_squared_error │   mean_absolute_error │      r2 │   error │
╞═════════╪══════════╪══════════════════════╪═══════════════════════╪═════════╪═════════╡
│ train   │ 793.5813 │             793.5813 │               25.1056 │ -0.0183 │ 24.5648 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ vali    │ 856.6062 │             856.6062 │               26.0861 │ -0.0207 │ 25.4445 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ test    │ 848.6282 │             848.6282 │               25.8354 │ -0.0184 │ 24.9674 │
╘═════════╧══════════╧══════════════════════╧═══════════════════════╧═════════╧═════════╛
Last improvement of loss on combined happened 4 epochs ago
INFO:ludwig.models.model:Last improvement of loss on combined happened 4 epochs ago

INFO:ludwig.models.model:

Epoch    5
INFO:ludwig.models.model:
Epoch    5
Training: 100%|██████████| 5/5 [00:00<00:00,  7.74it/s]
Evaluation train: 100%|██████████| 5/5 [00:00<00:00, 26.22it/s]
Evaluation vali : 100%|██████████| 1/1 [00:00<00:00, 40.37it/s]
Evaluation test : 100%|██████████| 2/2 [00:00<00:00, 34.16it/s]
Took 0.9471s
INFO:ludwig.models.model:Took 0.9471s
╒═════════╤══════════╤══════════════════════╤═══════════════════════╤═════════╤═════════╕
│ Score   │     loss │   mean_squared_error │   mean_absolute_error │      r2 │   error │
╞═════════╪══════════╪══════════════════════╪═══════════════════════╪═════════╪═════════╡
│ train   │ 915.9881 │             915.9881 │               27.5607 │ -0.0235 │ 27.3258 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ vali    │ 975.1177 │             975.1177 │               28.3510 │ -0.0257 │ 28.0044 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ test    │ 977.1770 │             977.1770 │               28.2331 │ -0.0233 │ 27.7329 │
╘═════════╧══════════╧══════════════════════╧═══════════════════════╧═════════╧═════════╛
INFO:ludwig.models.model:╒═════════╤══════════╤══════════════════════╤═══════════════════════╤═════════╤═════════╕
│ Score   │     loss │   mean_squared_error │   mean_absolute_error │      r2 │   error │
╞═════════╪══════════╪══════════════════════╪═══════════════════════╪═════════╪═════════╡
│ train   │ 915.9881 │             915.9881 │               27.5607 │ -0.0235 │ 27.3258 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ vali    │ 975.1177 │             975.1177 │               28.3510 │ -0.0257 │ 28.0044 │
├─────────┼──────────┼──────────────────────┼───────────────────────┼─────────┼─────────┤
│ test    │ 977.1770 │             977.1770 │               28.2331 │ -0.0233 │ 27.7329 │
╘═════════╧══════════╧══════════════════════╧═══════════════════════╧═════════╧═════════╛
Last improvement of loss on combined happened 5 epochs ago
INFO:ludwig.models.model:Last improvement of loss on combined happened 5 epochs ago

EARLY STOPPING due to lack of validation improvement, it has been 5 epochs since last validation accuracy improvement.
```

**Environment (please complete the following information):**
 - OS: Google Colab
- Python version 3.6
- Ludwig version 0.2

**Additional context**
This was working correctly in 0.1.2

**Comment by w4nderlust** [comment]
Thank you, we are looking into it.

**Comment by msaisumanth** [comment]
@danielduckworth Could you please provide the model definition? Also, could you please clarify if these two were completely separate experiments or not? To be precise, is the second experiment resuming from the first model (saved to disk) or sth like that?

**Comment by danielduckworth** [comment]
Sorry for the delayed response. I will test with some generated data and see if I can produce a re-producible error and then submit the data and model definition.

**Comment by w4nderlust** [comment]
@danielduckworth thank you. Will keep the issue open, let us know.

**Comment by w4nderlust** [comment]
I tried to replicate the issue with a known dataset and wasn't able to see any difference in behaviour between training with csv and df. Actually, if you look at the code, the only difference between the two is that is you call with csv internally the csv is loaded ( https://github.com/uber/ludwig/blob/master/ludwig/data/preprocessing.py#L60 ) so without a reproducible example it is really difficult for me to identify the issue.
I also suggest to install ludwig from master as, if there is an issue it could have been already solved.
Closing it for now, feel free to reopen if you have a reproducible example.

**Comment by danielduckworth** [comment]
No worries @w4nderlust. I'll keep an eye on it next time I'm working with Ludwig. The only thing I can think of is that the model definition used multiple text inputs and one numerical output. But yeah, master may have fixed it so I'll try that also.
